File: npc_sim/simulation/spatial_grid.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Dictionary-bucketed uniform spatial grid for fast NPC radius queries."""

from __future__ import annotations
import math
from abc import ABC, abstractmethod
from npc_sim.core.sim_vector3 import SimVector3
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from npc_sim.npc.npc import NPC


class ISpatialGrid(ABC):
    @abstractmethod
    def insert(self, npc) -> None: ...
    @abstractmethod
    def remove(self, npc) -> None: ...
    @abstractmethod
    def update(self, npc) -> None: ...
    @abstractmethod
    def clear(self) -> None: ...
    @abstractmethod
    def query_radius(self, center: SimVector3, radius: float) -> list: ...


class DictionaryGrid(ISpatialGrid):
    """O(1) insert/remove, O(k) radius query where k = NPCs in candidate cells."""

    def __init__(self, cell_size: float = 50.0):
        self._cell_size = max(1.0, cell_size)
        self._cells: dict[int, list] = {}
        self._index: dict[str, tuple[int, int]] = {}

    def insert(self, npc) -> None:
        cx, cy = self._cell(npc.position)
        self._get_or_create(cx, cy).append(npc)
        self._index[npc.identity.npc_id] = (cx, cy)

    def remove(self, npc) -> None:
        cell = self._index.get(npc.identity.npc_id)
        if cell is None:
            return
        key = self._key(cell[0], cell[1])
        bucket = self._cells.get(key)
        if bucket:
            try:
                bucket.remove(npc)
            except ValueError:
                pass
        self._index.pop(npc.identity.npc_id, None)

    def update(self, npc) -> None:
        old = self._index.get(npc.identity.npc_id)
        if old:
            ncx, ncy = self._cell(npc.position)
            if ncx == old[0] and ncy == old[1]:
                return
            key = self._key(old[0], old[1])
            bucket = self._cells.get(key)
            if bucket:
                try:
                    bucket.remove(npc)
                except ValueError:
                    pass
        self.insert(npc)

    def clear(self) -> None:
        self._cells.clear()
        self._index.clear()

    def query_radius(self, center: SimVector3, radius: float) -> list:
        cell_r = int(math.ceil(radius / self._cell_size))
        cx, cy = self._cell(center)
        result = []
        sqr_r = radius * radius

        for dx in range(-cell_r, cell_r + 1):
            for dy in range(-cell_r, cell_r + 1):
                key = self._key(cx + dx, cy + dy)
                bucket = self._cells.get(key)
                if not bucket:
                    continue
                for npc in bucket:
                    if SimVector3.sqr_distance(center, npc.position) <= sqr_r:
                        result.append(npc)
        return result

    def _cell(self, pos: SimVector3) -> tuple[int, int]:
        return (int(math.floor(pos.x / self._cell_size)),
                int(math.floor(pos.z / self._cell_size)))

    def _get_or_create(self, cx: int, cy: int) -> list:
        key = self._key(cx, cy)
        if key not in self._cells:
            self._cells[key] = []
        return self._cells[key]

    @staticmethod
    def _key(x: int, y: int) -> int:
        return (x << 32) | (y & 0xFFFFFFFF)

File: Runtime/Simulation/SpatialGrid.cs
using System;
using System.Collections.Generic;

namespace ForgeProject.Sim
{
    // ─── Spatial Partition Abstraction ────────────────────────────────────────────

    /// <summary>
    /// Abstracts the spatial indexing strategy so <see cref="StimulusDispatcher"/>
    /// and <see cref="SimWorldAdapter"/> remain decoupled from the partition algorithm.
    /// Swap in a quad-tree or Unity NativeArray grid without touching dispatch logic.
    /// </summary>
    public interface ISpatialGrid
    {
        void Insert(NPC npc);
        void Remove(NPC npc);
        void Update(NPC npc);
        void Clear();
        IReadOnlyList<NPC> QueryRadius(SimVector3 center, float radius);
    }

    /// <summary>
    /// Default spatial grid: dictionary-bucketed uniform grid.
    /// O(1) insert/remove, O(k) radius query where k = NPCs in candidate cells.
    /// Good for dense urban placement. Cell size should be ≥ max sensor radius.
    /// </summary>
    public sealed class DictionaryGrid : ISpatialGrid
    {
        private readonly float _cellSize;
        private readonly Dictionary<long, List<NPC>> _cells = new();
        private readonly Dictionary<string, (int cx, int cy)> _index  = new();

        public DictionaryGrid(float cellSize = 50f)
        {
            _cellSize = MathF.Max(1f, cellSize);
        }

        public void Insert(NPC npc)
        {
            var (cx, cy) = Cell(npc.Position);
            GetOrCreate(cx, cy).Add(npc);
            _index[npc.Identity.NpcId] = (cx, cy);
        }

        public void Remove(NPC npc)
        {
            if (!_index.TryGetValue(npc.Identity.NpcId, out var cell)) return;
            var key = Key(cell.cx, cell.cy);
            if (_cells.TryGetValue(key, out var list)) list.Remove(npc);
            _index.Remove(npc.Identity.NpcId);
        }

        public void Update(NPC npc)
        {
            if (_index.TryGetValue(npc.Identity.NpcId, out var old))
            {
                var (ncx, ncy) = Cell(npc.Position);
                if (ncx == old.cx && ncy == old.cy) return;  // same cell, skip
                var oldKey = Key(old.cx, old.cy);
                if (_cells.TryGetValue(oldKey, out var list)) list.Remove(npc);
            }
            Insert(npc);
        }

        public void Clear() { _cells.Clear(); _index.Clear(); }

        public IReadOnlyList<NPC> QueryRadius(SimVector3 center, float radius)
        {
            int cellR = (int)MathF.Ceiling(radius / _cellSize);
            var (cx, cy) = Cell(center);
            var result   = new List<NPC>();
            float sqrR   = radius * radius;

            for (int dx = -cellR; dx <= cellR; dx++)
            for (int dy = -cellR; dy <= cellR; dy++)
            {
                var key = Key(cx + dx, cy + dy);
                if (!_cells.TryGetValue(key, out var bucket)) continue;
                foreach (var npc in bucket)
                    if (SimVector3.SqrDistance(center, npc.Position) <= sqrR)
                        result.Add(npc);
            }
            return result;
        }

        private (int cx, int cy) Cell(SimVector3 pos)
            => ((int)MathF.Floor(pos.X / _cellSize), (int)MathF.Floor(pos.Z / _cellSize));

        private List<NPC> GetOrCreate(int cx, int cy)
        {
            var key = Key(cx, cy);
            if (!_cells.TryGetValue(key, out var list))
                _cells[key] = list = new List<NPC>();
            return list;
        }

        private static long Key(int x, int y) => ((long)x << 32) | (uint)y;
    }
}
